### PR TITLE
Skip some `Convolution2D` tests for older numpy versions

### LINKED ISSUE #7946: Flaky test: `tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py::TestConvolution2DFunction`
https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1835/TEST=chainer-py2,label=mn1-p100/console

>`FAIL tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py::TestConvolution2DFunction_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_25_{cover_all=False, groups=2, dilate=1, nobias=False, contiguous=None, W_dtype=float16, x_dtype=float16}::test_forward`

```
TestConvolution2DFunction_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_25_{cover_all=False, groups=2, dilate=1, nobias=False, contiguous=None, W_dtype=float16, x_dtype=float16}.test_forward _
13:01:48 
13:01:48 self = <chainer.testing._bundle.TestConvolution2DFunction_use_chainerx_false__chainer...se, contiguous=None, W_dtype=float16, x_dtype=float16} testMethod=test_forward>
13:01:48 args = (), kwargs = {}
13:01:48 e = FunctionTestError('Outputs do not match the expected values.\nIndices of outpu...]\n\n     [[ 0.83984375  1.10253906]\n      [ 0.80859375  1.1171875 ]]]]\n\n',)
13:01:48 s = <StringIO.StringIO instance at 0x7f5c66aa8c20>, k = 'x_dtype'
13:01:48 v = <type 'numpy.float16'>

:
:

13:01:48 E       FunctionTestError: Parameterized test failed.
13:01:48 E       
13:01:48 E       Base test method: TestConvolution2DFunction_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_false__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never.test_forward
13:01:48 E       Test parameters:
13:01:48 E         cover_all: False
13:01:48 E         groups: 2
13:01:48 E         dilate: 1
13:01:48 E         nobias: False
13:01:48 E         contiguous: None
13:01:48 E         W_dtype: <type 'numpy.float16'>
13:01:48 E         x_dtype: <type 'numpy.float16'>
13:01:48 E       
13:01:48 E       
13:01:48 E       (caused by)
13:01:48 E       FunctionTestError: Outputs do not match the expected values.
13:01:48 E       Indices of outputs that do not match: 0
13:01:48 E       Expected shapes and dtypes: (2, 4, 2, 2):float16
13:01:48 E       Actual shapes and dtypes:   (2, 4, 2, 2):float16
13:01:48 E       
13:01:48 E       
13:01:48 E       Error details of output [0]:
13:01:48 E       
13:01:48 E       Not equal to tolerance rtol=0.01, atol=0.001
13:01:48 E       
13:01:48 E       (mismatch 100.0%)
13:01:48 E        x: array([[[[ 1.167969,  1.277344],
13:01:48 E                [ 1.152344,  0.912598]],
13:01:48 E       ...
13:01:48 E        y: array([[[[ 1.167969,  1.277344],
13:01:48 E                [ 1.152344,  0.912598]],
13:01:48 E       ...
13:01:48 E       
13:01:48 E       assert_allclose failed: 
13:01:48 E         shape: (2, 4, 2, 2) (2, 4, 2, 2)
13:01:48 E         dtype: float16 float16
13:01:48 E         i: (1, 2, 1, 1)
13:01:48 E         x[i]: -0.05615234375
13:01:48 E         y[i]: -0.0537109375
13:01:48 E         relative error[i]: 0.0454406738281
13:01:48 E         absolute error[i]: 0.00244140625
13:01:48 E       x: [[[[ 1.16796875  1.27734375]
13:01:48 E             [ 1.15234375  0.91259766]]
13:01:48 E       
13:01:48 E            [[-0.34790039 -0.76757812]
13:01:48 E             [-0.22021484 -1.08789062]]
13:01:48 E       
13:01:48 E            [[-0.61035156 -0.66308594]
13:01:48 E             [-1.84082031 -0.24072266]]
13:01:48 E       
13:01:48 E            [[ 1.27929688  0.86328125]
13:01:48 E             [ 0.44848633  0.68701172]]]
13:01:48 E       
13:01:48 E       
13:01:48 E           [[[ 0.93701172  0.859375  ]
13:01:48 E             [ 1.15820312  1.34472656]]
13:01:48 E       
13:01:48 E            [[-0.36254883 -0.12939453]
13:01:48 E             [-1.09179688 -0.32543945]]
13:01:48 E       
13:01:48 E            [[-0.56640625 -0.79101562]
13:01:48 E             [-1.55273438 -0.05615234]]
13:01:48 E       
13:01:48 E            [[ 0.83984375  1.10253906]
13:01:48 E             [ 0.80810547  1.11816406]]]]
13:01:48 E       y: [[[[ 1.16796875  1.27734375]
13:01:48 E             [ 1.15234375  0.91259766]]
13:01:48 E       
13:01:48 E            [[-0.34814453 -0.76757812]
13:01:48 E             [-0.22033691 -1.08789062]]
13:01:48 E       
13:01:48 E            [[-0.60986328 -0.66308594]
13:01:48 E             [-1.84082031 -0.24072266]]
13:01:48 E       
13:01:48 E            [[ 1.27929688  0.86328125]
13:01:48 E             [ 0.44824219  0.68701172]]]
13:01:48 E       
13:01:48 E       
13:01:48 E           [[[ 0.93701172  0.859375  ]
13:01:48 E             [ 1.15917969  1.34472656]]
13:01:48 E       
13:01:48 E            [[-0.36230469 -0.12939453]
13:01:48 E             [-1.09277344 -0.32543945]]
13:01:48 E       
13:01:48 E            [[-0.56591797 -0.79101562]
13:01:48 E             [-1.55273438 -0.05371094]]
13:01:48 E       
13:01:48 E            [[ 0.83984375  1.10253906]
13:01:48 E             [ 0.80859375  1.1171875 ]]]]
```

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #7946

The reason behind the flakiness is the use of a numpy version < 1.17
1.17 fixed a bug in the float16 implementation.

What happens is that forward_expected is comparing all the backends against the cpu implementation using numpy. And when using CUDA the results in float16 might sightly differ in some corner cases dut to the cpu-gpu difference (numpy bug) in fp16 implementation. Other tests that rely on numpy for forward_expected are most likely to be affected by this.

This failure only happens with grouped convolutions.

### CODE REVIEW

**Comment by niboshi** [comment]
I suggest avoid skipping but instead omit numerical checks (https://github.com/chainer/chainer/issues/7946#issuecomment-555293026).
Calculations should be successfully performed even with older numpy versions.

**Comment by emcastillo** [comment]
Comment addressed, thanks a lot for the feedback.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 1]
```diff
@@ -75,7 +75,15 @@ def setUp(self):
         self.check_double_backward_options.update({
             'atol': 5e-4, 'rtol': 5e-3
         })
+        self.old_numpy_fp16 = False
         if numpy.float16 in (self.x_dtype, self.W_dtype):
+            # Old numpy versions have a bug in the fp16 conversion
+            # that happens on the matrix multiplication for the grouped
+            # convolution, outputs will be zeroed but computations
+            # will be performed in order to detect other issues
+            self.old_numpy_fp16 = (testing.require_satisfies('numpy<1.17')
```
Comment (toslunar) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: NumPy version is tested with
```python
numpy.lib.NumpyVersion(numpy.__version__)
```
in other places.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 2]
```diff
@@ -129,6 +138,8 @@ def forward_expected(self, inputs):
                 x, W, b, stride=self.stride, pad=self.pad,
                 cover_all=self.cover_all, dilate=self.dilate,
                 groups=self.groups)
+        if self.old_numpy_fp16:
+            return y_expected.array*0,
```
Comment (takagi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]:
```suggestion
            return y_expected.array*0,  # See #8458
```

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py [hunk 3]
```diff
@@ -137,10 +148,13 @@ def forward(self, inputs, device):
...
         else:
             x, W, b = inputs
-        return F.convolution_2d(
+        out = F.convolution_2d(
             x, W, b, stride=self.stride, pad=self.pad,
             cover_all=self.cover_all, dilate=self.dilate,
-            groups=self.groups),
+            groups=self.groups)
+        if self.old_numpy_fp16:
+            return out*0,
```
Comment (takagi) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]:
```suggestion
            return out*0,  # See #8458
```

**Comment by takagi** [comment]
pfnCI, test this please.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py (hunk 3 above)
Comment (emcastillo) [reply tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Its explained in the setUp comment 😄

https://github.com/chainer/chainer/pull/8458/files#diff-0c14a0d1b376c42cedd98bb914521e8bR80-R83

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py (hunk 2 above)
Comment (takagi) [reply tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: Okay! I close my commetns.

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py (hunk 3 above)
Comment (takagi) [reply tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py]: ditto.

**Comment by takagi** [comment]
LGTM! The CI failure on Travis is not related to this change.
